pkg/mcs/resource_manager/server/server.go

```diff
@@ -0,0 +1,306 @@
+	"google.golang.org/grpc"
+)
+
+const tcp = "tcp"
+
+// Server is the resource manager server, and it implements bs.Server.
+// nolint
+type Server struct {
+	cfg *Config
+	mux cmux.CMux
+	// Server state. 0 is not serving, 1 is serving.
+	isServing   int64
+	ctx         context.Context
+	name        string
+	backendUrls []*url.URL
+
+	etcdClient *clientv3.Client
+	httpClient *http.Client
+
+	grpcServer *grpc.Server
+	httpServer *http.Server
+	service    *Service
+
+	serverLoopWg sync.WaitGroup
+	// Callback functions for different stages
+	// startCallbacks will be called after the server is started.
+	startCallbacks []func()
+	// primaryCallbacks will be called after the server becomes leader.
+	primaryCallbacks []func(context.Context)
+}
+
+// Name returns the unique etcd Name for this server in etcd cluster.
+func (s *Server) Name() string {
+	return s.name
+}
+
+// Context returns the context.
+func (s *Server) Context() context.Context {
+	return s.ctx
+}
+
+// Run runs the pd server.
+func (s *Server) Run() error {
+	if err := s.initClient(); err != nil {
+		return err
+	}
+	if err := s.startServer(); err != nil {
+		return err
+	}
+	return s.mux.Serve()
+}
+
+// Close closes the server.
+func (s *Server) Close() {
+	if !atomic.CompareAndSwapInt64(&s.isServing, 1, 0) {
+		// server is already closed
+		return
+	}
+
+	log.Info("closing server")
+
+	s.grpcServer.Stop()
+	s.httpServer.Shutdown(s.ctx)
+	s.serverLoopWg.Wait()
+
+	if s.etcdClient != nil {
+		if err := s.etcdClient.Close(); err != nil {
+			log.Error("close etcd client meet error", errs.ZapError(errs.ErrCloseEtcdClient, err))
+		}
+	}
+
+	if s.httpClient != nil {
+		s.httpClient.CloseIdleConnections()
+	}
+}
+
+// GetClient returns builtin etcd client.
+func (s *Server) GetClient() *clientv3.Client {
+	return s.etcdClient
+}
+
+// GetHTTPClient returns builtin http client.
+func (s *Server) GetHTTPClient() *http.Client {
+	return s.httpClient
+}
+
+// AddStartCallback adds a callback in the startServer phase.
+func (s *Server) AddStartCallback(callbacks ...func()) {
+	s.startCallbacks = append(s.startCallbacks, callbacks...)
+}
+
+// IsServing returns whether the server is the leader, if there is embedded etcd, or the primary otherwise.
+func (s *Server) IsServing() bool {
+	// TODO: implement this function with primary.
+	return s.isServing == 1
+}
+
+// AddServiceReadyCallback adds the callback function when the server becomes the leader, if there is embedded etcd, or the primary otherwise.
+func (s *Server) AddServiceReadyCallback(callbacks ...func(context.Context)) {
+	s.primaryCallbacks = append(s.primaryCallbacks, callbacks...)
+}
+
+func (s *Server) initClient() error {
+	tlsConfig, err := s.cfg.Security.ToTLSConfig()
+	if err != nil {
+		return err
+	}
+	endpoints := strings.Split(s.cfg.BackendEndpoints, ",")
+	for _, endpoint := range endpoints {
+		e, err := url.Parse(endpoint)
+		if err != nil {
+			return err
+		}
+		s.backendUrls = append(s.backendUrls, e)
+	}
```

Comment:
Shall we replace line 150 ~ 157 with the following code?

		u, err := types.NewURLs(strings.Split(s.cfg.BackendEndpoints, ","))
		if err != nil {
                        return err
		}
            s.backendUrls = []url.URL(u)

Reply:
shall we define the type of s.backendUrls as []url.URL instead of []*url.URL? Any reason for defining it as []*url.URL?

Reply:
After changing the type of s.backendUrls to []url.URL , the code is (if no bug)
	u, err := types.NewURLs(strings.Split(s.cfg.BackendEndpoints, ","))
	if err != nil {
		return err
	}
	s.backendUrls = []url.URL(u)
	s.etcdClient, s.httpClient, err = etcdutil.CreateClients(tlsConfig, s.backendUrls)

Reply:
BTW, types.NewURLs() already checked:

	if len(all) == 0 {
		return nil, errors.New("no valid URLs given")
	}

Reply:
good idea